### ISSUE TITLE
Possibility of @ConditionalOnProperty support on spring-native

### ISSUE THREAD

**Issue by jgslima** [spring-projects-experimental/spring-native#1613]
This is more a question (I followed the guideline to open [a question in stackoverflow](https://stackoverflow.com/questions/72093864/possibility-of-conditionalonproperty-support-on-spring-native) first, but as it had no answer I am posting the question here).

I am trying as hard as I can to understand exactly the meaning, effects and all reasons of having to make all the preparation of the BeanFactory at build time.

I suppose it would not make sense to try to build an application ready to evaluate at runtime all the possible conditions of all AutoConfigurations of spring-boot. Countless hints would be needed, and the footprint would be unnecessarily high.

However, I wonder whether would be possible for spring-native to create an application capable of evaluating, at runtime, _at least some specific conditions_, which would be specified by the developer at build time. Among all conditions of org.springframework.boot.autoconfigure.condition, supporting "just" `@ConditionalOnProperty` would be a game changer.

The point is about applications that may be installed in distinct environments. The company I work for makes several software products, which may be installed not only in distinct clouds, but may also integrate with different pieces of middlewares (for instance different message brokers). We use `@ConditionalOnProperty` heavily throughout all applications. Without its support, I cannot see how we could migrate to spring-native.

The idea would be the developer to indicate somehow (maybe as hints) specific classes anoted with `@ConditionalOnProperty` which should be honored at runtime.

Of course, the developer would be willing to pay the price of having a footprint/image with possible unused code.

Apologies if this idea does not even makes sense with the current principles of spring-native.

**Comment by bclozel** [comment]
No worries @jgslima, sorry for not spotting your question on StackOverflow. As the answer is really subjective, I'll post my opinion here.

Right now all `org.springframework.context.annotation.Condition` are processed during the AOT phase and we get a bean factory with bean definitions as a result. In theory, we could indeed expand the condition contract to have an optional AOT behavior that would generate source code to allow for more flexibility at runtime. Note that this would probably require some more flexibility: application developers would need to configure namespaces so that some properties are considered as fixed at build time, or not.

From what we're seeing in a typical Spring Boot application right now, `@ConditionalOnProperty` is mostly tied to `*.enabled` properties which can import large bits of infrastructure, if the classpath allows it. This could make the image size bigger.

To summarize:
* we're now focused on moving Spring Native features to Spring Framework as first class support
* AOT conditions support could be done in Spring Framework 6, but we don't have plans for this right now
* depending on how it's implemented, this could impact the native image size significantly

**Comment by jgslima** [comment]
Thank you for the quick an detailed answer!

**Comment by jgslima** [comment]
As a matter of fact, as I mentioned:

> I am trying as hard as I can to understand exactly the (...) reasons of having to make all the preparation of the BeanFactory at build time.

I had forgot about the fact that one of the principles of the GraalVM native image is to not have lazy class loading. Everything shipped is loaded in memory.

I approached native images interested exactly in low memory consumption. So, my own request of having runtime condition evaluation would harm this, because we would load to memory unused classes. Opposed to what a regular JVM does.

**Comment by snicoll** [comment]
There's not much we can do here I am afraid so I am going to close it. We will revisit this topic as Brian mentioned.
